plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -1,80 +1,213 @@
+-- We could maybe do this even more efficiently, but Seq should get the job done.
+-- | The order in which terms get evaluated, along with their purities.
+newtype EvalOrder tyname name uni fun a = EvalOrder (Seq.Seq (EvalTerm tyname name uni fun a))
+
+-- | Get the evaluation order as a list of 'EvalTerm's.
+unEvalOrder :: EvalOrder tyname name uni fun a -> [EvalTerm tyname name uni fun a]
+unEvalOrder (EvalOrder ts) = toList ts
+
+evalThis :: EvalTerm tyname name uni fun a -> EvalOrder tyname name uni fun a
+evalThis tm = EvalOrder (Seq.singleton tm)
+
+instance PrettyBy config (Term tyname name uni fun a)
+  => PrettyBy config (EvalOrder tyname name uni fun a) where
+  prettyBy config eo = vsep $ fmap (prettyBy config) (unEvalOrder eo)
+
+instance Semigroup (EvalOrder tyname name uni fun a) where
+  -- If we have discovered that we don't know where we're going, then no point recording after that
+  EvalOrder l@(_ Seq.:|> Unknown) <> EvalOrder _ = EvalOrder l
+  EvalOrder l <> EvalOrder r                     = EvalOrder (l <> r)
+
+instance Monoid (EvalOrder tyname name uni fun a) where
+  mempty = EvalOrder mempty
+
+{- | Given a term, return the order in which it and its sub-terms will be evaluated.
+
+This aims to be a sound under-approximation: if we don't know, we just say 'Unknown'.
+Typically there will be a sequence of terms that we do know, which will terminate
+in 'Unknown' once we do something like call a function.
+
+This makes some assumptions about the evaluator, in particular about the order in
+which we evaluate sub-terms, but these match the current evaluator and we are not
+planning on changing it.
 -}
+termEvaluationOrder
+  :: forall tyname name uni fun a
+  . ToBuiltinMeaning uni fun
+  => BuiltinSemanticsVariant fun
+  -> (name -> Strictness)
+  -> Term tyname name uni fun a
+  -> EvalOrder tyname name uni fun a
+termEvaluationOrder semvar varStrictness = goTerm
+    where
+      goTerm :: Term tyname name uni fun a -> EvalOrder tyname name uni fun a
+      goTerm = \case
+        t@(Let _ NonRec bs b) ->
+          goBindings (NE.toList bs) -- first the bindings, in order
+          <> goTerm b -- then the body
+          <> evalThis (EvalTerm Pure t) -- then the whole term
+        (Let _ Rec _ _)   ->
+          -- Hard to know what gets evaluated first in a recursive let-binding,
+          -- just give up
+          evalThis Unknown
+
+        -- If we can view as a builtin application, then handle that specially
+        (splitApplication -> (Builtin a fun, args)) -> goBuiltinApp a fun args
+        -- We could handle functions and type abstractions with *known* bodies
+        -- here. But there's not much point: beta reduction will immediately
+        -- turn those into let-bindings, which we do see through already.
+        t@(Apply _ fun arg)    ->
+          goTerm fun -- first the function
+          <> goTerm arg -- then the arg
+          <> evalThis (EvalTerm Pure t) -- then the whole term
```

Comment:
So, this would be relevant if e.g. you wanted to do the work-free check. You'd say something like "it's okay if the evaluation order includes work-free terms like `LamAbs`, but not if it includes e.g. actual applications". So you need to see the application step in the order. Plus it corresponds to what actually happens in the machine and the reduction rules: we evaluate the function, then the argument, then we do something for the application.

Reply:
This is in the code now - we need those nodes otherwise we can't say why e.g. `[(lam x x) (con integer 1)]` is not work-free.